### PR TITLE
Calendar export

### PR DESCRIPTION
**Pull request by miggland**
This is related to issue #3031 and is a replacement for PR #3391.

New URLs:
```
http://inventree.example.org/api/order/calendar/purchase-order/calendar.ics
http://inventree.example.org/api/order/calendar/sales-order/calendar.ics
```

Parameter:
`?include_completed=True`
to include the completed orders in the calendar. By default, these are not included.

<a href="https://gitpod.io/#https://github.com/inventree/InvenTree/pull/3858"></a>

### CODE REVIEW

**Comment by miggland** [comment]
When starting this container, I'm getting an error with a setting.

Any ideas what's wrong? This doesn't fail if I have the container running, then apply the new code..

I'm running these:
```
> docker-compose build
> docker-compose run inventree-dev-server invoke install
> docker-compose run inventree-dev-server invoke update
```
```
Installing plugin packages from '/home/inventree/data/plugins.txt'
Backing up InvenTree database...
Traceback (most recent call last):
  File "/home/inventree/data/env/lib/python3.9/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
psycopg2.errors.UndefinedTable: relation "common_inventreesetting" does not exist
LINE 1: ...ng"."value", "common_inventreesetting"."key" FROM "common_in...
                                                             ^


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/inventree/InvenTree/manage.py", line 23, in <module>
    execute_from_command_line(sys.argv)
  File "/home/inventree/data/env/lib/python3.9/site-packages/django/core/management/__init__.py", line 419, in execute_from_command_line
    utility.execute()
  File "/home/inventree/data/env/lib/python3.9/site-packages/django/core/management/__init__.py", line 413, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/home/inventree/data/env/lib/python3.9/site-packages/django/core/management/base.py", line 354, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/home/inventree/data/env/lib/python3.9/site-packages/django/core/management/base.py", line 393, in execute
    self.check()
  File "/home/inventree/data/env/lib/python3.9/site-packages/django/core/management/base.py", line 419, in check
    all_issues = checks.run_checks(
  File "/home/inventree/data/env/lib/python3.9/site-packages/django/core/checks/registry.py", line 76, in run_checks
    new_errors = check(app_configs=app_configs, databases=databases)
  File "/home/inventree/data/env/lib/python3.9/site-packages/django/core/checks/urls.py", line 40, in check_url_namespaces_unique
    all_namespaces = _load_all_namespaces(resolver)
  File "/home/inventree/data/env/lib/python3.9/site-packages/django/core/checks/urls.py", line 57, in _load_all_namespaces
    url_patterns = getattr(resolver, 'url_patterns', [])
  File "/home/inventree/data/env/lib/python3.9/site-packages/django/utils/functional.py", line 48, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
  File "/home/inventree/data/env/lib/python3.9/site-packages/django/urls/resolvers.py", line 602, in url_patterns
    patterns = getattr(self.urlconf_module, "urlpatterns", self.urlconf_module)
  File "/home/inventree/data/env/lib/python3.9/site-packages/django/utils/functional.py", line 48, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
  File "/home/inventree/data/env/lib/python3.9/site-packages/django/urls/resolvers.py", line 595, in urlconf_module
    return import_module(self.urlconf_name)
  File "/usr/local/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 850, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "/home/inventree/InvenTree/InvenTree/urls.py", line 22, in <module>
    from order.api import order_api_urls
  File "/home/inventree/InvenTree/order/api.py", line 1128, in <module>
    class OrderCalendarExport(ICalFeed):
  File "/home/inventree/InvenTree/order/api.py", line 1136, in OrderCalendarExport
    instance_url = InvenTreeSetting.get_setting("INVENTREE_BASE_URL").replace("http://", "").replace("https://", "")
  File "/home/inventree/InvenTree/common/models.py", line 383, in get_setting
    setting = cls.get_setting_object(key, **kwargs)
  File "/home/inventree/InvenTree/common/models.py", line 335, in get_setting_object
    setting = settings.filter(**filters).first()
  File "/home/inventree/data/env/lib/python3.9/site-packages/django/db/models/query.py", line 674, in first
    for obj in (self if self.ordered else self.order_by('pk'))[:1]:
  File "/home/inventree/data/env/lib/python3.9/site-packages/django/db/models/query.py", line 280, in __iter__
    self._fetch_all()
  File "/home/inventree/data/env/lib/python3.9/site-packages/django/db/models/query.py", line 1324, in _fetch_all
    self._result_cache = list(self._iterable_class(self))
  File "/home/inventree/data/env/lib/python3.9/site-packages/django/db/models/query.py", line 51, in __iter__
    results = compiler.execute_sql(chunked_fetch=self.chunked_fetch, chunk_size=self.chunk_size)
  File "/home/inventree/data/env/lib/python3.9/site-packages/django/db/models/sql/compiler.py", line 1175, in execute_sql
    cursor.execute(sql, params)
  File "/home/inventree/data/env/lib/python3.9/site-packages/django/db/backends/utils.py", line 98, in execute
    return super().execute(sql, params)
  File "/home/inventree/data/env/lib/python3.9/site-packages/django/db/backends/utils.py", line 66, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/home/inventree/data/env/lib/python3.9/site-packages/django/db/backends/utils.py", line 75, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/home/inventree/data/env/lib/python3.9/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/home/inventree/data/env/lib/python3.9/site-packages/django/db/utils.py", line 90, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/home/inventree/data/env/lib/python3.9/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
django.db.utils.ProgrammingError: relation "common_inventreesetting" does not exist
LINE 1: ...ng"."value", "common_inventreesetting"."key" FROM "common_in...
                                                             ^

ERROR: 1
```

**Review by matmair — APPROVED** [review]

**Comment by matmair** [comment]
Interesting @miggland ... nothing jumps into the eye directly. I will try to find the problem on the weekend

File: InvenTree/order/api.py
Comment (matmair) [inline InvenTree/order/api.py:1164]: This seems to be the wrong setting

File: InvenTree/order/api.py
Comment (matmair) [inline InvenTree/order/api.py:1290]: Suggestion: use instance url as referenced above

File: requirements.txt
Comment (matmair) [inline requirements.txt:173]: There are multiple dependency updates here, maybe issue them in a separate PR

**Review by matmair — CHANGES_REQUESTED** [review]

File: InvenTree/order/api.py
Comment (miggland) [reply InvenTree/order/api.py:1164]: Doh..

Too much copy-paste. Hey, it worked though ;) I'll correct it.

File: requirements.txt
Comment (miggland) [reply requirements.txt:173]: Yeah that was because of a merge, I guess.. not intended. I'll check and remove the wrong ones though.

File: InvenTree/order/api.py
Comment (miggland) [reply InvenTree/order/api.py]: This is fixed.

**Review by matmair — APPROVED** [review]
looks good!

**Review by matmair — APPROVED** [review]

File: requirements.txt
Comment (miggland) [reply requirements.txt]: OK, I've reduced the number of changes here. However, I don't understand how this works.

How do I get this right? If I run `invoke setup-dev` in my repo, outside of Docker, I get updates to the pre-commit yaml file which have to be committed before I can continue. Is that expected?

Also, I keep getting changes to the header of the requirements.txt file because I have python3.10 instead of python3.9.

For now, I've forced it to be as you expected via Github - but it cost me a lot of time, and it's not something I really want to be spending time on.

**Comment by matmair** [comment]
@miggland am I seeing correctly that the tests do not export any events? Could rewrite at least one of them to check if iterating over the items causes any breakage?
Also - if possible - a test for the custom auth would be great. You will need to explicitly log out the user before making the test call. Example code for that:
https://github.com/inventree/InvenTree/blob/ff400106d7f9a60e33e2b64b28ff252d9b8ac021/InvenTree/InvenTree/test_api.py#L84-L93C4

File: requirements.txt
Comment (matmair) [reply requirements.txt]: @miggland that is not expected behaviour. It could be that it changes your env the first time you run 'setup-dev' as it needs to pull the latest devtools and runs pip install for that.
Changes in the dependencies can have unforeseen consequences and as we do not have 100% coverage it is better to not do them without some manual testing/canary beforehand. This PR is probably one of the last before the new stable release so better no experiments.

**Comment by miggland** [comment]
> Am I seeing correctly that the tests do not export any events? Could rewrite at least one of them to check if iterating over the items causes any breakage?

Sorry, I don't understand what you mean. Do you mean that the test doesn't check that there are actually orders, as in number_orders >= 1?

**Comment by matmair** [comment]
> Sorry, I don't understand what you mean. Do you mean that the test doesn't check that there are actually orders, as in number_orders >= 1?

Yes. Take a look at the coverage: https://coveralls.io/builds/54550055. Especially the point where events are [hard-coded to be zero](https://coveralls.io/builds/54550055/source?filename=InvenTree%2Forder%2Ftest_api.py#L388) and that none of the [item attributes](https://coveralls.io/builds/54550055/source?filename=InvenTree%2Forder%2Fapi.py#L1262) is defined

File: requirements.txt
Comment (miggland) [reply requirements.txt]: > Changes in the dependencies can have unforeseen consequences and as we do not have 100% coverage it is better to not do them without some manual testing/canary beforehand. This PR is probably one of the last before the new stable release so better no experiments.

Yes, I have no problem with that.

But how the *)(F)* do I set up my system to allow me to work without causing problems? You can't expect everyone to have exactly the same version of every package installed on their system, *outside* of Docker - if that's what you want, it should be in a controlled image.

**Comment by matmair** [comment]
This might be because there is no fixture loaded. Easiest would be to just add an event at beginning of the tests

**Comment by miggland** [comment]
> This might be because there is no fixture loaded. Easiest would be to just add an event at beginning of the tests

@matmair I looked in the web interface, and saw orders which should have worked - but you're right, there were no orders present. I've improved the tests to make sure there are sufficient orders before doing the test.

Trying with the unauthenticated user as well, but I'm not sure I'll get that to work yet..

**Comment by miggland** [comment]
> Also - if possible - a test for the custom auth would be great. You will need to explicitly log out the user before making the test call. Example code for that:
> https://github.com/inventree/InvenTree/blob/ff400106d7f9a60e33e2b64b28ff252d9b8ac021/InvenTree/InvenTree/test_api.py#L84-L93C4

@matmair: I'm not getting this to work. How do I explicitly log out the user?
What I tried:
```
self.client.credentials(HTTP_AUTHORIZATION="")
response = self.client.get(url)
```

**Comment by miggland** [comment]
> > Also - if possible - a test for the custom auth would be great. You will need to explicitly log out the user before making the test call. Example code for that:
> > https://github.com/inventree/InvenTree/blob/ff400106d7f9a60e33e2b64b28ff252d9b8ac021/InvenTree/InvenTree/test_api.py#L84-L93C4
> 
> @matmair: I'm not getting this to work. How do I explicitly log out the user? What I tried:
> 
> ```
> self.client.credentials(HTTP_AUTHORIZATION="")
> response = self.client.get(url)
> ```

I've added a test for this, so by now most of the code should be covered. I also merged the current master into this branch, so that it's based on the newest code.

**Review by matmair — APPROVED** [review]

**Comment by SchrodingersGat** [comment]
@miggland I'll try to get to this soon. Can you give me a brief summary of how I should test it?

**Comment by matmair** [comment]
@miggland I added the missing test for successful basic auth

**Comment by miggland** [comment]
Thanks @matmair!

@SchrodingersGat:
For basic usage: I've added this to the docs as well (see https://github.com/inventree/inventree-docs/pull/399/files)

The URLs are:
`http://inventree.example.org/api/order/purchase-order/calendar.ics`
`http://inventree.example.org/api/order/sales-order/calendar.ics`

Add `?include_completed=True` in order to see completed orders also.

You can either download the file through a browser, or use a client such as Thunderbird to view the calendar. Login with normal credentials in this case (username/password).

**Comment by SchrodingersGat** [comment]
@miggland can you please update the API version in `api_version.py` :)

**Comment by SchrodingersGat** [comment]
@miggland having issue getting it working locally:

image

Any ideas?

**Comment by miggland** [comment]
Yes. Sorry about that.. I made a mistake in the documentation, and copied that over. I'll correct it.

**Comment by SchrodingersGat** [comment]
Looks well implemented, I'd be happy to merge in (if you're finished) once conflicts are fixed :)

**Comment by miggland** [comment]
Fixed the conflict (since you merged the other PR so quickly :)).

I'm done with this - the same logic should be implemented for Build orders as well, but I'd say leave that to the future for now..

**Comment by SchrodingersGat** [comment]
Happy to leave build orders, thanks for your work here :)

**Pull request merged**
